### PR TITLE
Initial version of Memoizer tests

### PR DESCRIPTION
**Pull request by melissalinkert**
See https://trello.com/c/x2nmObSk/191-bfmemo-change-detection

As mentioned in today's formats meeting, I wouldn't necessarily expect this to be merged as-is - further discussion may be warranted.  Certainly feel free to exclude if needed, as I likely won't be able to do much further with this until 18 June.

Main points are:

* Memoizer now has a ```setReadOnly(boolean)``` method so that it can be prevented from deleting/overwriting existing memo files
* ```cachegen``` command line tool can be used to generate memo files for a single file, directory, or list of files (enumerated in a text file) in a specified directory
* ```gen-cache``` target in ```components/test-suite``` can be used to generate memo files in the directory specified by ```testng.cacheDirectory``` (using existing directory scanning logic)
* ```test-automated``` target will automatically test reading the memo files in ```testng.cacheDirectory```, if defined
* ```test-memoizer``` target in ```components/test-suite``` only tests memo files, and will check what is in ```testng.cacheDirectory```, if defined (it otherwise requires the same properties as ```test-automated```)
* ```gen-file-list``` target in ```components/test-suite``` saves the list of scanned files to ```testng.file-list``` (requires ```testng.directory``` and ```testng.configDirectory```)

I wouldn't expect any immediate impact on builds, as ```testng.cacheDirectory``` should be optional.

An initial test would be to pick a directory under curated and do something like:

```
# generate memo files:
$ ant -Dtestng.directory=/path/to/curated/dir/ -Dtestng.configDirectory=/path/to/data_repo_config/curated/dir/ -Dtestng.cacheDirectory=/path/to/cache/ gen-cache
# verify that tests pass and there is no logging to indicate missing memo files:
$ ant -Dtestng.directory=/path/to/curated/dir/ -Dtestng.configDirectory=/path/to/data_repo_config/curated/dir/ -Dtestng.cacheDirectory=/path/to/cache/ test-memoizer
# ... add a new field to a reader and rebuild ...
# all tests should now fail:
$ ant -Dtestng.directory=/path/to/curated/dir/ -Dtestng.configDirectory=/path/to/data_repo_config/curated/dir/ -Dtestng.cacheDirectory=/path/to/cache/ test-memoizer
# confirm that the memo files were not altered since originally written by gen-cache:
$ ls -al /path/to/cache/
```

### CODE REVIEW

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [BIOFORMATS-DEV-merge-push#1404](https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-push/1404/). See the [console output](https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-push/1404/consoleText) for more details.
Possible conflicts:
  - PR #3140 rleigh-codelibre 'Support for unconfigured tests'
    - components/test-suite/build.xml
  - Upstream changes
    - components/test-suite/build.xml

--conflicts

**Comment by sbesson** [comment]
Started a first round of testing of this PR by trying to run the `gen-cache` target on a format folder:

```
$ docker run --rm --name BIOFORMATS-cache -v /uod/idr/repos/curated/biorad-gel:/opt/data -v /scratch/data_repo_config/curated/biorad-gel:/opt/config -v /scratch/caches/curated/biorad-gel/:/opt/cache --entrypoint "/opt/ant/bin/ant" openmicroscopy/bioformats:develop gen-cache -Dtestng.directory=/opt/data -Dtestng.configDirectory=/opt/config -Dtestng.cacheDirectory=/opt/cache -Dtestng.memory=15g -Dtestng.threadCount=12
Buildfile: /opt/bioformats/components/test-suite/build.xml
     [echo] isSnapshot = true

mvn-init:

init-dependencies:
     [echo] Loading dependency paths from file: /opt/bioformats/components/test-suite/build/compile-dependencies.xml

init-dependencies:
     [echo] Loading dependency paths from file: /opt/bioformats/components/test-suite/build/test-dependencies.xml

init-dependencies:
     [echo] Loading dependency paths from file: /opt/bioformats/components/test-suite/build/runtime-dependencies.xml

init-title:
     [echo] ----------=========== bio-formats-testing-framework ===========----------

init-timestamp:

init:

copy-resources:

compile:

gen-cache:
   [testng] [TestNG] [Error] 
   [testng] An error occurred while instantiating class loci.tests.testng.FormatReaderTestFactory: org/slf4j/LoggerFactory
   [testng] The tests failed.

BUILD FAILED
/opt/bioformats/components/test-suite/build.xml:154: if=failedTest

Total time: 0 seconds
```

while 

```
$ docker run --rm --name BIOFORMATS-test -v /uod/idr/repos/curated/biorad-gel:/opt/data -v /scratch/data_repo_config/curated/biorad-gel:/opt/config -v /scratch/caches/curated/biorad-gel/:/opt/cache --entrypoint "/opt/ant/bin/ant" openmicroscopy/bioformats:develop test-automated -Dtestng.directory=/opt/data -Dtestng.configDirectory=/opt/config -Dtestng.memory=15g -Dtestng.threadCount=12
```

works as expected (as shown by the daily test-repo job). The other new targets `test-memoizer`, `gen-file-list` also fail in the `FormatReaderTestFactory` so I suspect this is specific to the newly added targets but did not have time to investigate further.

**Comment by melissalinkert** [comment]
7b49a8e should fix the build failures noted above.

**Comment by sbesson** [comment]
Testing of the last version with 

```
docker run --rm --name BIOFORMATS-memo -v /uod/idr/repos/curated/biorad-gel:/opt/data -v /scratch/data_repo_config/curated/biorad-gel:/opt/config -v /scratch/caches/curated/biorad-gel/:/opt/cache --entrypoint /opt/ant/bin/ant openmicroscopy/bioformats:develop gen-cache -Dtestng.directory=/opt/data -Dtestng.configDirectory=/opt/config -Dtestng.cacheDirectory=/opt/cache -Dtestng.memory=15g -Dtestng.threadCount=12
```

resulted in a lot of exception since the `/opt/cache` directory is mounted as `root` and not writeable by the user running the test targets. Pushed an extra commit to create `/opt/data` as a location owned by the `bf` user.

File: components/bio-formats-tools/src/loci/formats/tools/GenerateCache.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      System.out.println();
+      System.out.println("If '-list' is specified, then 'fileOrDir' is a text file with one file per line.");
+      return;
+    }
+
+    boolean fileList = args.length >= 3 && args[0].equals("-list");
+    String input = args[args.length - 2];
+    String outputDir = args[args.length - 1];
+
+    Memoizer reader = new Memoizer(0, new File(outputDir));
```
[the remark this answers is not in the document]
Comment (melissalinkert) [reply components/bio-formats-tools/src/loci/formats/tools/GenerateCache.java]: Almost certainly makes sense to add options for extra wrappers, either to explicitly specify the reader stack or do what ```showinf``` allows via ```-stitch```, ```-separate```, etc.  It should then be easy enough to match OMERO's reader stack.

File: components/formats-bsd/src/loci/formats/Memoizer.java [hunk 2]
```diff
@@ -607,6 +607,19 @@ public void setVersionChecking(boolean version) {
...
+  /**
+   * Set whether a memo file should be saved if a valid file is not present.
+   *
+   * If {@code false} (default), then a memo file will be saved.  This may
+   * cause an existing memo file to be overwritten.
+   *
+   * If {@code true}, then a memo file will not be saved.  This effectively
+   * makes the current Memoizer read-only.
+   */
+  public void setReadOnly(boolean skip) {
```
[the remark this answers is not in the document]
Comment (melissalinkert) [reply components/formats-bsd/src/loci/formats/Memoizer.java]: If just renaming this method to ```skipSave(boolean)``` helps clarify, then that's probably easiest.  The intention was not for ```setReadOnly(true)``` to imply that the cache directory permissions should be something like 555, but rather to tell Memoizer not to touch a (probably legitimately) read/write directory.

**Comment by melissalinkert** [comment]
Can definitely rename the Memoizer API as noted above.  For the second point, I think we want ```cachegen``` no matter what, but having ```gen-cache``` call ```cachegen``` would probably be fine.  I can push a commit for the API rename later today; is the rest blocking this PR or could it be done as a follow-up?

**Comment by sbesson** [comment]
@melissalinkert I think renaming the Memoizer API to have some containment the scope/expectations and match the internal field name is sufficient. From my side, none of the other comments are blockers to getting this PR merged. As stated above, most of this is targetted for internal test consumption and we probably want to announce them as such in the release notes /cc @dgault

**Comment by melissalinkert** [comment]
f7102d4 renames the new Memoizer API and https://trello.com/c/yEGoa4Gb/255-cache-generation-follow-up captures remaining tasks for a future PR.

**Pull request merged**